src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,450 @@
+// the form:
+// ```Datalog
+//      Q_NAME = query ASSERTION(<ARGS>) ?
+// ```
+//
+// is translated into an assertion with just one argument, and this
+// assertion is made an output, so we can simply check the CSV that
+// souffle emits as in:
+// ```Datalog
+//      Q_NAME("dummy_var") :- grounded("dummy_var"), ASSERTION(<ARGS>).
+//      grounded("dummy_var").
+//      .output Q_NAME
+// ```
+
+class LoweringToDatalogPass {
+ public:
+  static DLIRProgram Lower(const Program& auth_logic_program) {
+    return LoweringToDatalogPass().ProgToDLIR(auth_logic_program);
+  }
+
+ private:
+  LoweringToDatalogPass() : fresh_var_count_(0) {}
+
+  // This returns a fresh variable name. Fresh variables are introduced in a
+  // few places during the translation.
+  std::string FreshVar() {
+    CHECK(fresh_var_count_ < std::numeric_limits<uint64_t>::max());
+    return "x__" + std::to_string(++fresh_var_count_);
+  }
+
+  // This function makes a copy of a predicate that prepends the name with
+  // another string and prepends the list of arguments with additional
+  // arguments. This is used in a few places in the translation, for
+  // example, to translate "X says blah(args)" into "says_blah(X, args)".
+  Predicate PushOntoPred(std::string modifier,
+                         std::vector<std::string> new_args,
+                         Predicate predicate) {
+    std::string new_name = modifier + predicate.name();
+    new_args.insert(new_args.end(),
+                    std::make_move_iterator(predicate.args().begin()),
+                    std::make_move_iterator(predicate.args().end()));
+    Sign sign_copy = predicate.sign();
+    return Predicate(new_name, new_args, sign_copy);
+  }
+
+  Predicate PushPrin(std::string modifier, Principal principal,
+                     Predicate predicate) {
+    return PushOntoPred(std::move(modifier), {principal.name()}, predicate);
+  }
+
+  Predicate AttributeToDLIR(Attribute attribute) {
+    // If attribute is `X pred(args...)` the following predicate is
+    // `pred(X, args...)`
+    return PushPrin(std::string(""), attribute.principal(),
+                    attribute.predicate());
+  }
+
+  // Translating an attribute results in two objects:
+  // - A predicate, this is the main result of the translation
+  // - An additional rule that allows attributes to be passed with canActAs
+  // This function is for generating the additional rule.
+  // This takes both a speaker and the actual attribute as argument because
+  // the speaker is needed to generate the additional rule. Note that all base
+  // facts have a speaker either explicitly or implicitly:
+  // - If it appears on the LHS of an assertion, it explicitly has a speaker
+  // - If it appears on the RHS of an assertion, it behaves semantically
+  // like it has the same speaker as the head of the assertion.
+  DLIRAssertion SpokenAttributeToDLIR(Principal speaker, Attribute attribute) {
```

Comment:
Functions with large implementation should go to C++ file.

Reply:
done. (also for a few others).